compat/mingw.c

```diff
@@ -1328,162 +1357,96 @@ static wchar_t *make_environment_block(char **deltaenv)
 	 * as a function that returns a pointer to a mostly static table.
 	 * Grab the pointer and cache it for the duration of our loop.
 	 */
-	const wchar_t *my_wenviron = GetEnvironmentStringsW();
-	const wchar_t *p;
+	const wchar_t *wenv = GetEnvironmentStringsW(), *p;
+	size_t delta_size = 0, size = 1; /* for extra NUL at the end */
 
-	/*
-	 * Internally, we create normal 'C' arrays of strings (pointing
-	 * into the blocks) to help with some of the de-dup work.
-	 */
-	wchar_t **wptrs_ins = NULL;
-	wchar_t **wptrs_del = NULL;
-	wchar_t *wblock_ins = NULL;
-	wchar_t *wblock_del = NULL;
-	wchar_t *wend_ins;
-	wchar_t *wend_del;
-	wchar_t *w_ins;
-	wchar_t *w_del;
-
-	int maxlen = 0;
-	int nr_delta = 0;
-	int nr_delta_ins = 0;
-	int nr_delta_del = 0;
-	int nr_wenv = 0;
-	int j, k, k_ins, k_del;
+	wchar_t **array = NULL;
+	size_t alloc = 0, nr = 0, i;
+
+	const char *p2;
+	wchar_t *wdeltaenv;
+
+	wchar_t *result, *p3;
 
 	/*
-	 * Count the number of inserts and deletes in the deltaenv list.
-	 * Allocate 'C' arrays for inserts and deletes.
-	 * Also estimate the block size of our results.
+	 * If there is no deltaenv to apply, simply return a copy
 	 */
-	if (deltaenv && deltaenv[0] && *deltaenv[0]) {
-		for (k = 0; deltaenv && deltaenv[k] && *deltaenv[k]; k++) {
-			if (strchr(deltaenv[k], '=') == NULL)
-				nr_delta_del++;
-			else {
-				maxlen += strlen(deltaenv[k]) + 1;
-				nr_delta_ins++;
-			}
+	if (!deltaenv || !*deltaenv) {
+		for (p = wenv; p && *p; ) {
+			size_t s = wcslen(p) + 1;
+			size += s;
+			p += s;
 		}
 
-		if (nr_delta_ins)
-			wptrs_ins = (wchar_t**)calloc(nr_delta_ins + 1, sizeof(wchar_t*));
-		if (nr_delta_del)
-			wptrs_del = (wchar_t**)calloc(nr_delta_del + 1, sizeof(wchar_t*));
-
-		nr_delta = nr_delta_ins + nr_delta_del;
-	}
-	for (p = my_wenviron; p && *p; ) {
-		size_t len = wcslen(p) + 1;
-		maxlen += len;
-		p += len;
-		nr_wenv++;
+		ALLOC_ARRAY(result, size);
+		memcpy(result, wenv, size * sizeof(*wenv));
+		FreeEnvironmentStringsW(wenv);
+		return result;
 	}
-	maxlen++;
 
 	/*
-	 * Allocate blocks for inserted and deleted items.
-	 * The individual pointers in the 'C' arrays will point into here.
-	 * We will use the wblock_ins as the final result.
+	 * If there is a deltaenv, let's accumulate all keys into `array`,
+	 * sort them using the stable git_qsort() and then copy, skipping
+	 * duplicate keys
```

Comment:
Do we care about order within the deltaenv?  That is, if they pass a delete-then-set vs a set-then-delete, do we want different results?

Reply:
Yes, that's why I need the stable sort, and that's why `wenvcmp()` does not simply `wcscmp()` (or `_wcsicmp()`).

Reply:
I mean if they pass in deltaenv:

    KEY
    KEY=foo

vs

    KEY=foo
    KEY

I would expect different results when de-duping deltaenv.  In the first case, the result has KEY=foo, but in the second case the result does not have KEY, right?

I'm not seeing how the wenvcmp() preserves this, but i'll look again.

Reply:
Yes, that is my reading about the order, too.

So how does the stable sort together with the complicated way `wenvcmp()` ensures to compare *only* the keys do that?

Well, first of all, `wenvcmp()` tries to match only the keys, i.e. `KEY=foo` is identical to `KEY`. If we used something like `wcscmp()` directly, `KEY` would always sort *before* `KEY=foo`. Not what we want in your second example.

Second, the stable sort ensures that keys considered identical by the comparison function preserve their relative order. That is, if you add `KEY=foo`, then a couple of other variables, and then `KEY`, and if your comparison function considers `KEY=foo` to be equivalent to `KEY`, then the sorted output will still have `KEY=foo` before `KEY`, still.

The [final copying loop](https://github.com/git-for-windows/git/pull/1304/files/5a28d2e6c4742be4186e361392388f89d5c83e3c..1e902ee76e6ca57516e3a15f2f55a4ac911f28bd#diff-a0c7aa297d8da1149ce761c31dc59328R1423) then makes sure that entries are skipped that are to-be-deleted entries, and also entries are skipped that are overridden by the respective next entry (for which I still need to convert the case-sensitive comparison, too).